Userland/Libraries/LibJS/Bytecode/Op.cpp

```diff
@@ -182,6 +182,54 @@ ThrowCompletionOr<void> NewArray::execute_impl(Bytecode::Interpreter& interprete
     return {};
 }
 
+ThrowCompletionOr<void> ConcatArray::execute_impl(Bytecode::Interpreter& interpreter) const
+{
+    // Note: This OpCode is used to construct array literals containing at least one spread element,
```

Comment:
eh?
I still don't understand why you insist on shoving all the iteration into an op, this changes the order of evaluation _drastically_, you would note in the very section you mentioned below, that evaluation of an argument is delayed until the spread is _completely_ evaluated, you evaluate all the arguments first, then perform the iterator dances necessary - this is _not_ correct given something like
```js
function foo() {
    console.log(arguments);
}

function* bar() {
    foo(
        yield 1,
       (yield 2)...,
       yield 3
    );
}

function* baz() {
    yield 1;
    console.log("lol");
    yield 2;
}

let g = bar();
g.next(); // get to the first arg
console.log("aaa");
g.next(1); // first arg = 1
console.log("oh no");
g.next(baz()); // second arg = 1, third arg = 2, random "lol" printed
console.log("whoops");
g.next(4); // fourth arg = 4
g.next();
```
Now the expectation is to get
```
aaa
oh no
lol
whoops
Arguments [1, 1, 2, 4]
```
your method of evaluation would yield the `lol` at the _end_ of the argument list evaluation.

As I've probably said before, you really only need to generate code like this, which matches the spec more closely anyway:
```js
// given:
foo(a, b, c..., d);
```
```
// generate
@start:
    // start with `a, b`:
    (a)
    Store $1
    (b)
    Store $2
    NewArray $1..$2
    Store $3
    // then add in stuff from c, manually iterating over it
    (c)
    GetIterator
    Store $4
    Jump @l
@l:
    Load $4
    IteratorNext
    Store $5
    IteratorResultValue
    AppendToArray $3 // new op
    Load $5
    IteratorResultDone
    JumpConditional true:@done false:@l
@done:
    (d)
    AppendToArray $3
    // resolve and call 'foo' now.
    (foo)
    Store $6
    Load $3
    Call callee: $6 this: ... arguments: acc
```

Reply:
On a closer inspection, your code does seem to be evaluating the arguments in the correct order, it's just really confusing to me I suppose.

I'm not entirely sure if the way you've implemented it has the correct behaviour all the time, but it looks alright...I'd still suggest just generating a loop instead though.

Reply:
I think generating a loop would be more complex than the current form and I would want this merged, to get some other stuff PR'd that would cause merge-conficts with this.
If you/we find a fundamentally better solution in the future, be it for optimizability or correctness, we can always change it later

Reply:
in all honesty, I'd rather keep the opcodes as simple as possible, but :shrug: